pkg/pgmodel/upgrade_tests/upgrade_test.go

```diff
@@ -0,0 +1,366 @@
+			t.Fatal(err)
+		}
+		defer db.Close()
+		f(container, db, tmpDir, t)
+	})
+}
+
+func migrateToVersion(t testing.TB, connectURL string, version string, commitHash string) {
+	migratePool, err := pgxpool.Connect(context.Background(), connectURL)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer migratePool.Close()
+	err = Migrate(migratePool, pgmodel.VersionInfo{Version: version, CommitHash: commitHash})
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+var testDirCounter = 0
+
+func TestDevUpgrades(t *testing.T) {
+	// we test that upgrading from the previous version gives the correct output
+	// by induction, this property should hold true for any chain of versions
+	prevVersion := semver.MustParse(version.Version)
+	toPreviousVersion(&prevVersion)
+
+	var upgradedDbInfo dbInfo
+	withNewDB(t, cleanImage, prevVersion.String(), *testDatabase,
+		func(container testcontainers.Container, db *pgxpool.Pool, tmpDir string, _ testing.TB) {
+			ingestor, err := pgmodel.NewPgxIngestor(db)
+			if err != nil {
+				t.Fatalf("error connecting to DB: %v", err)
+			}
+			data := []prompb.TimeSeries{
+				{
+					Labels: []prompb.Label{
+						{Name: MetricNameLabelName, Value: "test"},
+						{Name: "test", Value: "test"},
+					},
+					Samples: []prompb.Sample{
+						{Timestamp: 1, Value: 0.1},
+						{Timestamp: 2, Value: 0.2},
+					},
+				},
+				{
+					Labels: []prompb.Label{
+						{Name: MetricNameLabelName, Value: "test2"},
+						{Name: "foo", Value: "bar"},
+					},
+					Samples: []prompb.Sample{
+						{Timestamp: 4, Value: 2.2},
+					},
+				},
+			}
+			_, err = ingestor.Ingest(data, &prompb.WriteRequest{})
+			upgradedDbInfo = getDbInfo(t, container, tmpDir, db)
+		})
+
+	var pristineDbInfo dbInfo
+	withNewDB(t, cleanImage, version.Version, *testDatabase,
+		func(container testcontainers.Container, db *pgxpool.Pool, tmpDir string, _ testing.TB) {
+			ingestor, err := pgmodel.NewPgxIngestor(db)
+			if err != nil {
+				t.Fatalf("error connecting to DB: %v", err)
+			}
+			data := []prompb.TimeSeries{
+				{
+					Labels: []prompb.Label{
+						{Name: MetricNameLabelName, Value: "test"},
+						{Name: "test", Value: "test"},
+					},
+					Samples: []prompb.Sample{
+						{Timestamp: 1, Value: 0.1},
+						{Timestamp: 2, Value: 0.2},
+					},
+				},
+				{
+					Labels: []prompb.Label{
+						{Name: MetricNameLabelName, Value: "test2"},
+						{Name: "foo", Value: "bar"},
+					},
+					Samples: []prompb.Sample{
+						{Timestamp: 4, Value: 2.2},
+					},
+				},
+			}
+			_, err = ingestor.Ingest(data, &prompb.WriteRequest{})
+			pristineDbInfo = getDbInfo(t, container, tmpDir, db)
+		})
+
+	if !reflect.DeepEqual(upgradedDbInfo, pristineDbInfo) {
+		t.Errorf("upgrade differences")
+	}
+}
+
+func toPreviousVersion(version *semver.Version) {
+	if len(version.Pre) > 0 {
+		lastPreDigit := &version.Pre[len(version.Pre)-1]
+		if lastPreDigit.VersionNum > 0 {
+			lastPreDigit.VersionNum -= 1
+			return
+		}
+
+		if len(version.Pre) > 2 {
```

Comment:
why are we skipping 2 here. Why not just go backward in Pre slice order and decrease any version that is numeric?